### ISSUE TITLE
[Bug]: [scale] The replica information is not updated after scale in milvus cluster

### ISSUE THREAD

**Issue by ThreadDao** [milvus-io/milvus#16705]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: master-20220428-ed51ffd5
- Deployment mode(standalone or cluster): cluster
- SDK version(e.g. pymilvus v2.0.0rc2): pymilvus 2.1.0.dev33
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

1. Deploy milvus cluster with **5** querynodes
2. Create a collection with 2 shards
3. Insert 30000 entities 10 times and flush
4. Load collection with **2** replicas
5. Scale in querynode from 5 to **1** during search
6. Server logs:  
[mic_logs.tar.gz](https://github.com/milvus-io/milvus/files/8580490/mic_logs.tar.gz)

**Search while scale in get the exception:**
```
[2022-04-28 15:13:40,312 - DEBUG - ci_test]: (api_request)  : [Collection.search] args: [[[0.06528075807724913, 0.036847462680379896, 0.09246272970289168, 0.0733426841048508, 0.0032477713105877053, 0.1088067113101843, 0.09888046153216196, 0.1276629245803541, 0.04755330957279652, 0.11608363789248426, 0.04223285911195517, 0.09363487353611741, 0.10181181418322184, 0.11431535899387318, 0.0......, kwargs: {} (api_request.py:55)
[2022-04-28 15:13:45,743 - ERROR - ci_test]: Traceback (most recent call last):
  File "/Users/nausicca/projects/milvus/tests/python_client/utils/api_request.py", line 22, in inner_wrapper
    res = func(*args, **kwargs)
  File "/Users/nausicca/projects/milvus/tests/python_client/utils/api_request.py", line 56, in api_request
    return func(*arg, **kwargs)
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/orm/collection.py", line 672, in search
    res = conn.search(self._name, data, anns_field, param, limit, expr,
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/decorators.py", line 56, in handler
    raise e
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/decorators.py", line 41, in handler
    return func(self, *args, **kwargs)
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/decorators.py", line 74, in handler
    raise e
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/decorators.py", line 70, in handler
    return func(*args, **kwargs)
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/decorators.py", line 21, in handler
    return func(self, *args, **kwargs)
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/client/grpc_handler.py", line 481, in search
    return self._execute_search_requests(requests, timeout, **_kwargs)
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/decorators.py", line 74, in handler
    raise e
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/decorators.py", line 70, in handler
    return func(*args, **kwargs)
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/client/grpc_handler.py", line 428, in _execute_search_requests
    raise pre_err
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/client/grpc_handler.py", line 419, in _execute_search_requests
    raise MilvusException(response.status.error_code, response.status.reason)
pymilvus.exceptions.MilvusException: <MilvusException: (code=1, message=Invalid shard leader)>
 (api_request.py:35)
```
**Get replicas after scaling:**
```
[2022-04-28 15:29:04,319 - DEBUG - ci_test]: Replica groups:
- Group: <group_id:432833742676164610>, <group_nodes:(4, 6, 3)>, <shards:[Shard: <channel_name:by-dev-rootcoord-dml_3_432833812734410753v1>, <shard_leader:6>, <shard_nodes:[6, 4]>, Shard: <channel_name:by-dev-rootcoord-dml_2_432833812734410753v0>, <shard_leader:3>, <shard_nodes:[3, 4]>]>
- Group: <group_id:432833742676164611>, <group_nodes:(7, 1)>, <shards:[Shard: <channel_name:by-dev-rootcoord-dml_2_432833812734410753v0>, <shard_leader:7>, <shard_nodes:[7, 1]>, Shard: <channel_name:by-dev-rootcoord-dml_3_432833812734410753v1>, <shard_leader:1>, <shard_nodes:[1, 7]>]> (test_query_node_scale.py:299)
```
**Get query segment info after scaling:**
```
[2022-04-28 15:30:41,494 - DEBUG - ci_test]: (api_request)  : [get_query_segment_info] args: ['scale_out_T6wldi9C', 20, 'default'], kwargs: {} (api_request.py:55)
[2022-04-28 15:30:41,517 - ERROR - ci_test]: Traceback (most recent call last):
  File "/Users/nausicca/projects/milvus/tests/python_client/utils/api_request.py", line 22, in inner_wrapper
    res = func(*args, **kwargs)
  File "/Users/nausicca/projects/milvus/tests/python_client/utils/api_request.py", line 56, in api_request
    return func(*arg, **kwargs)
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/orm/utility.py", line 580, in get_query_segment_info
    return _get_connection(using).get_query_segment_info(collection_name, timeout=timeout)
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/decorators.py", line 56, in handler
    raise e
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/decorators.py", line 41, in handler
    return func(self, *args, **kwargs)
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/decorators.py", line 74, in handler
    raise e
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/decorators.py", line 70, in handler
    return func(*args, **kwargs)
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/client/grpc_handler.py", line 492, in get_query_segment_info
    raise MilvusException(status.error_code, status.reason)
pymilvus.exceptions.MilvusException: <MilvusException: (code=1, message=getSegmentInfo: queryNode 3 is offline)>
 (api_request.py:35)
[2022-04-28 15:30:41,519 - ERROR - ci_test]: (api_response) : <MilvusException: (code=1, message=getSegmentInfo: queryNode 3 is offline)> (api_request.py:36)
```
**Search after scaling get empty result:**
```
[2022-04-28 15:30:41,430 - DEBUG - ci_test]: (api_request)  : [Collection.search] args: [[[0.07910548859018086, 0.14085075658263346, 0.010471022238144344, 0.0663830840749841, 0.08056899516875833, 0.018545394890012577, 0.03158771387057227, 0.05252935883685282, 0.1274276619175271, 0.14668501881562926, 0.022711258981200296, 0.08037928482835252, 0.11763564428484789, 0.11941187823695741, 0......., kwargs: {} (api_request.py:55)
[2022-04-28 15:30:41,493 - DEBUG - ci_test]: (api_response) : []  (api_request.py:27)
```

### Expected Behavior

To be discussed

### Steps To Reproduce

```markdown
def test_scale_in_query_node_less_than_replicas(self):
        release_name = "mic-xxx"
        mic = MilvusOperator()
        host = mic.endpoint(release_name, constants.NAMESPACE).split(':')[0]

        try:
            connections.connect("default", host=host, port=19530)

            collection_w = ApiCollectionWrapper()
            collection_w.init_collection(name=cf.gen_unique_str("scale_out"), schema=cf.gen_default_collection_schema())

            # insert 10 sealed segments
            for i in range(5):
                df = cf.gen_default_dataframe_data(start=i * ct.default_nb)
                collection_w.insert(df)
                assert collection_w.num_entities == (i + 1) * ct.default_nb

            collection_w.load(replica_number=2)

            @counter
            def do_search():
                """ do search """
                search_res, is_succ = collection_w.search(cf.gen_vectors(1, ct.default_dim),
                                                          ct.default_float_vec_field_name,
                                                          ct.default_search_params, ct.default_limit)
                log.debug(search_res[0].ids)
                assert len(search_res[0].ids) == ct.default_limit
                return search_res, is_succ

            def loop_search():
                """ continuously search """
                while True:
                    do_search()

            t_search = threading.Thread(target=loop_search, args=(), daemon=True)
            t_search.start()

            # scale in
            mic.upgrade(release_name, {'spec.components.queryNode.replicas': 1}, constants.NAMESPACE)
            mic.wait_for_healthy(release_name, constants.NAMESPACE)
            wait_pods_ready(constants.NAMESPACE, f"app.kubernetes.io/instance={release_name}")
            log.debug("Scale in querynode success")

            time.sleep(100)
            log.debug("Scale in test finished")

            log.debug(f"total: {do_search.total}, succ: {do_search.succ}, fail: {do_search.fail}")
            assert do_search.fail == 0
            assert do_search.succ // do_search.total == 1

        except Exception as e:
            log.debug(str(e))

        finally:
            log.info(f'Test finished')
```

**Comment by yanliang567** [comment]
@xiaofan-luan I think we should design how Milvus behaves when querynodes shrink to less than replica number

/assign @xiaofan-luan

**Comment by ThreadDao** [comment]
Image `master-20220428-206b554d`
1. The search is always successful whether it is scaling or after scaling.
2. **After scaling, collection replica info is not updated (Before scale, load with 2 replica number, then scale in querynode from 3 to 1):**
```
[2022-04-28 16:49:04,118 - DEBUG - ci_test]: Replica groups:
- Group: <group_id:432835221106982915>, <group_nodes:(3, 5)>, <shards:[Shard: <channel_name:by-dev-rootcoord-dml_1_432835264517505025v1>, <shard_leader:5>, <shard_nodes:[5, 3]>, Shard: <channel_name:by-dev-rootcoord-dml_0_432835264517505025v0>, <shard_leader:3>, <shard_nodes:[3, 5]>]>
- Group: <group_id:432835221106982914>, <group_nodes:(1,)>, <shards:[Shard: <channel_name:by-dev-rootcoord-dml_0_432835264517505025v0>, <shard_leader:1>, <shard_nodes:[1]>, Shard: <channel_name:by-dev-rootcoord-dml_1_432835264517505025v1>, <shard_leader:1>, <shard_nodes:[1]>]> (test_query_node_scale.py:299)
```
3. After scaling, the query segment info is correct
```
[2022-04-28 16:49:14,313 - DEBUG - ci_test]: [segmentID: 432835269511348225
collectionID: 432835264517505025
partitionID: 432835264517505026
mem_size: 766612
num_rows: 1463
nodeID: 1
state: Sealed
, segmentID: 432835269511348226
collectionID: 432835264517505025
partitionID: 432835264517505026
mem_size: 805388
num_rows: 1537
nodeID: 1
state: Sealed
, segmentID: 432835265670938625
collectionID: 432835264517505025
partitionID: 432835264517505026
mem_size: 787572
num_rows: 1503
nodeID: 1
state: Sealed
, segmentID: 432835265670938626
collectionID: 432835264517505025
partitionID: 432835264517505026
mem_size: 784428
num_rows: 1497
nodeID: 1
state: Sealed
, segmentID: 432835267689447425
collectionID: 432835264517505025
partitionID: 432835264517505026
mem_size: 779188
num_rows: 1487
nodeID: 1
state: Sealed
, segmentID: 432835267689447426
collectionID: 432835264517505025
partitionID: 432835264517505026
mem_size: 792812
num_rows: 1513
nodeID: 1
state: Sealed
] 
```

**Comment by xiaofan-luan** [comment]
This is annoying case. Overall the group is currently a hard request so if querynode is < replica number the behavior is as expected since the balance policy doen't allow two replica share the same querynode.

It fall back to the issue that if one replica querynode crashes then all the search will be hang due to round robin the bad replica.

I highly suggest to change the replica group policy to best effort and the balance goal is to try our best to place different segment replicas to different nodes. What do you think @sunby

**Comment by ThreadDao** [comment]
Image `master-20220428-206b554d`
After cluster scale in, the querynode is < replica number:
1. query entity get empty result
2. release collection hangs
3. load a new collection hangs
If I scale out the cluster and increase the querynodes from 1 back to 3, loading a new collection hangs
[mic_logs.tar.gz](https://github.com/milvus-io/milvus/files/8584090/mic_logs.tar.gz)

**Comment by yah01** [comment]
Hi @ThreadDao,
This should be fixed by #16854

**Comment by ThreadDao** [comment]
> Hi @ThreadDao, This should be fixed by #16854

@yah01 There is still a little problem. First deploy 2 querynodes and load 2 replicas, then scale in querynode from 2 to 1, and finally scale back to 2. 
After a few minutes later, the replicas info is:
```
c.get_replicas('scale_in_LlRFsFMH')
Replica groups:
- Group: <group_id:433124509418782735>, <group_nodes:(2,)>, <shards:[Shard: <channel_name:by-dev-rootcoord-dml_2_433124801783005185v0>, <shard_leader:2>, <shard_nodes:[2]>, Shard: <channel_name:by-dev-rootcoord-dml_3_433124801783005185v1>, <shard_leader:2>, <shard_nodes:[2]>]>
- Group: <group_id:433124509418782734>, <group_nodes:(5, 10)>, <shards:[Shard: <channel_name:by-dev-rootcoord-dml_2_433124801783005185v0>, <shard_leader:5>, <shard_nodes:[5]>, Shard: <channel_name:by-dev-rootcoord-dml_3_433124801783005185v1>, <shard_leader:5>, <shard_nodes:[5]>]>
```
The query segment info:
```
c.get_query_segment_info('scale_in_LlRFsFMH')
RPC error: [get_query_segment_info], <MilvusException: (code=1, message=getSegmentInfo: queryNode 2 is offline)>, <Time:{'RPC start': '2022-05-11 11:42:02.034075', 'RPC error': '2022-05-11 11:42:02.107564'}>
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/client/stub.py", line 904, in get_query_segment_info
    return handler.get_query_segment_info(collection_name, timeout, **kwargs)
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/decorators.py", line 56, in handler
    raise e
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/decorators.py", line 41, in handler
    return func(self, *args, **kwargs)
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/decorators.py", line 74, in handler
    raise e
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/decorators.py", line 70, in handler
    return func(*args, **kwargs)
  File "/Users/nausicca/.virtualenvs/milvus/lib/python3.8/site-packages/pymilvus/client/grpc_handler.py", line 492, in get_query_segment_info
    raise MilvusException(status.error_code, status.reason)
pymilvus.exceptions.MilvusException: <MilvusException: (code=1, message=getSegmentInfo: queryNode 2 is offline)>
```
[mic_logs.tar.gz](https://github.com/milvus-io/milvus/files/8666613/mic_logs.tar.gz)

**Comment by yah01** [comment]
Hi @ThreadDao 
We don't support the feature of balancing between replicas now, and can't make sure the new node is added into the replica which has the less nodes because of the latency of removing offline nodes. So, this won't be solved for the current version.

**Comment by yah01** [comment]
@sunby @congqixia 
For the case that every replica has only 1 node, the LoadBalance will always fail, which leads to removing nodes for ReplicaInfo and SegmentInfo never been executed.

**Comment by yah01** [comment]
This will be fixed with the rebalance between replicas feature, which will be implemented after version 2.1

**Comment by ThreadDao** [comment]
current situation of image `master-20220520-d525e955`:
Scale querynode after loading 2 replicas: 2 -> 1 -> 2
1. After scaling querynode from 2 to 1, search successfully, get replicas with 2 querynodes, `- Group: <group_id:433328812105400357>, <group_nodes:(4,)>, <shards:[Shard: <channel_name:by-dev-rootcoord-dml_4_433329218436464641v0>, <shard_leader:4>, <shard_nodes:[4]>, Shard: <channel_name:by-dev-rootcoord-dml_5_433329218436464641v1>, <shard_leader:4>, <shard_nodes:[4]>]>
- Group: <group_id:433328812105400356>, <group_nodes:(10,)>, <shards:[Shard: <channel_name:by-dev-rootcoord-dml_4_433329218436464641v0>, <shard_leader:10>, <shard_nodes:[10]>, Shard: <channel_name:by-dev-rootcoord-dml_5_433329218436464641v1>, <shard_leader:10>, <shard_nodes:[10]>]> (test_query_node_scale.py:295)`
get query segment info `context deadline exceeded`

2. After scaling querynode from 1 back to 2, search successfully, segment info is correct, but the offline querynode-10 in the replica are not removed
```
Replica groups:
- Group: <group_id:433328812105400357>, <group_nodes:(4,)>, <shards:[Shard: <channel_name:by-dev-rootcoord-dml_4_433329218436464641v0>, <shard_leader:4>, <shard_nodes:[4]>, Shard: <channel_name:by-dev-rootcoord-dml_5_433329218436464641v1>, <shard_leader:4>, <shard_nodes:[4]>]>
- Group: <group_id:433328812105400356>, <group_nodes:(10, 11)>, <shards:[Shard: <channel_name:by-dev-rootcoord-dml_4_433329218436464641v0>, <shard_leader:11>, <shard_nodes:[11]>, Shard: <channel_name:by-dev-rootcoord-dml_5_433329218436464641v1>, <shard_leader:10>, <shard_nodes:[10, 11]>]>
```
```
[segmentID: 433329218920120321
collectionID: 433329218436464641
...
num_rows: 1503
nodeID: 4
state: Sealed
, segmentID: 433329219000860673
collectionID: 433329218436464641
...
num_rows: 1497
nodeID: 4
state: Sealed
, segmentID: 433329218920120321
collectionID: 433329218436464641
...
num_rows: 1503
nodeID: 11
state: Sealed
, segmentID: 433329219000860673
collectionID: 433329218436464641
...
num_rows: 1497
nodeID: 11
state: Sealed
]
```

**Comment by yah01** [comment]
@ThreadDao 
#17306 should fix this

**Comment by ThreadDao** [comment]
Fixed `master-20220602-48cf63a2`
